hikari/impl/stateful_guild_chunker.py

```diff
@@ -23,65 +23,298 @@
+            pass
+
+        if not self._missing_chunks:
+            # While it is tempting to just close the stream here, that'd lead to a type error being
+            # raised the next time they try to iterate over it.
+            assert self._registered_listener is not None
+            self._app.dispatcher.unsubscribe(shard_events.MemberChunkEvent, self._registered_listener)
+
+    async def __anext__(self) -> shard_events.MemberChunkEvent:
+        if not self._active:
+            raise TypeError("stream must be started with `await with` before entering it")
+
+        if self._queue.empty() and not self._missing_chunks and self._missing_chunks is not None:
+            raise StopAsyncIteration from None
+
+        try:
+            return await asyncio.wait_for(self._queue.get(), timeout=self._timeout)
+        except asyncio.TimeoutError:
+            raise StopAsyncIteration from None
+
+    async def open(self) -> None:
+        # as super().open will possibly flip self.started, we want to have its state when this was originally called.
+        started = self._active
+        await super().open()
+
+        if not started:
+            shard_id = snowflakes.calculate_shard_id(self._app, self._guild_id)
+            await self._app.shards[shard_id].request_guild_members(
+                guild=self._guild_id,
+                include_presences=self._include_presences,
+                query=self._query,
+                limit=self._limit,
+                users=self._users,
+                nonce=self._nonce,
+            )
+
 
-_LOGGER = logging.getLogger("hikari.guild_chunker")
+@attr.s(init=True, kw_only=True, hash=False, eq=True, slots=True, weakref_slot=False)
+class _TrackedRequests:
+    """`hikari.api.chunker.RequestInformation` implementation."""
+
+    __slots__: typing.Sequence[str]
+
+    average_chunk_size: typing.Optional[int] = attr.ib(default=None)
+    chunk_count: typing.Optional[int] = attr.ib(default=None)
+    guild_id: snowflakes.Snowflake = attr.ib()
+    last_received: typing.Optional[datetime.datetime] = attr.ib(default=None)
+    _mono_last_received: typing.Optional[int] = attr.ib(default=None)
+    missing_chunk_indexes: typing.Optional[typing.MutableSequence[int]] = attr.ib(default=None)
+    nonce: str = attr.ib()
+    not_found_ids: typing.MutableSequence[snowflakes.Snowflake] = attr.ib(factory=list)
+
+    def __copy__(self) -> _TrackedRequests:
+        chunks = attr_extensions.copy_attrs(self)
+        chunks.missing_chunk_indexes = (
+            list(self.missing_chunk_indexes) if self.missing_chunk_indexes is not None else None
+        )
+        chunks.not_found_ids = list(self.not_found_ids)
+        return chunks
+
+    @property
+    def is_complete(self) -> bool:
+        if self.received_chunks == self.chunk_count:
+            return True
+
+        return self._mono_last_received is not None and time.monotonic_ns() - self._mono_last_received > EXPIRY_TIME
```

Comment:
Do we need nano precision here though

Reply:
This was just to avoid floating point math, it's either get the nano second one as an integer or get the normal one as a float which still includes nano seconds in the implementations of python i've seen